dpnp/dpnp_iface_indexing.py

```diff
@@ -482,20 +563,113 @@ def fill_diagonal(x1, val, wrap=False):
 
+    Examples
+    --------
+    >>> import dpnp as np
+    >>> a = np.zeros((3, 3), dtype=int)
+    >>> np.fill_diagonal(a, 5)
+    >>> a
+    array([[5, 0, 0],
+           [0, 5, 0],
+           [0, 0, 5]])
+
+    The same function can operate on a 4-D array:
+
+    >>> a = np.zeros((3, 3, 3, 3), dtype=int)
+    >>> np.fill_diagonal(a, 4)
+
+    We only show a few blocks for clarity:
+
+    >>> a[0, 0]
+    array([[4, 0, 0],
+           [0, 0, 0],
+           [0, 0, 0]])
+    >>> a[1, 1]
+    array([[0, 0, 0],
+           [0, 4, 0],
+           [0, 0, 0]])
+    >>> a[2, 2]
+    array([[0, 0, 0],
+           [0, 0, 0],
+           [0, 0, 4]])
+
+    The wrap option affects only tall matrices:
+
+    >>> # tall matrices no wrap
+    >>> a = np.zeros((5, 3), dtype=int)
+    >>> np.fill_diagonal(a, 4)
+    >>> a
+    array([[4, 0, 0],
+           [0, 4, 0],
+           [0, 0, 4],
+           [0, 0, 0],
+           [0, 0, 0]])
+
+    >>> # tall matrices wrap
+    >>> a = np.zeros((5, 3), dtype=int)
+    >>> np.fill_diagonal(a, 4, wrap=True)
+    >>> a
+    array([[4, 0, 0],
+           [0, 4, 0],
+           [0, 0, 4],
+           [0, 0, 0],
+           [4, 0, 0]])
+
+    >>> # wide matrices
+    >>> a = np.zeros((3, 5), dtype=int)
+    >>> np.fill_diagonal(a, 4, wrap=True)
+    >>> a
+    array([[4, 0, 0, 0, 0],
+           [0, 4, 0, 0, 0],
+           [0, 0, 4, 0, 0]])
+
+    The anti-diagonal can be filled by reversing the order of elements
+    using either `dpnp.flipud` or `dpnp.fliplr`.
+
+    >>> a = np.zeros((3, 3), dtype=int)
+    >>> val = np.array([1, 2, 3])
+    >>> np.fill_diagonal(np.fliplr(a), val)  # Horizontal flip
+    >>> a
+    array([[0, 0, 1],
+           [0, 2, 0],
+           [3, 0, 0]])
+    >>> np.fill_diagonal(np.flipud(a), val)  # Vertical flip
+    >>> a
+    array([[0, 0, 3],
+           [0, 2, 0],
+           [1, 0, 0]])
+
     """
 
-    x1_desc = dpnp.get_dpnp_descriptor(
-        x1, copy_when_strides=False, copy_when_nondefault_queue=False
-    )
-    if x1_desc:
-        if not dpnp.isscalar(val):
-            pass
-        elif wrap:
-            pass
-        else:
-            return dpnp_fill_diagonal(x1_desc, val)
+    dpnp.check_supported_arrays_type(a, scalar_type=True, all_scalars=True)
 
-    return call_origin(numpy.fill_diagonal, x1, val, wrap, dpnp_inplace=True)
+    if a.ndim < 2:
+        raise ValueError("array must be at least 2-d")
+    end = a.size
+    if a.ndim == 2:
+        step = a.shape[1] + 1
+        if not wrap and a.shape[0] > a.shape[1]:
+            end = a.shape[1] * a.shape[1]
+    else:
+        sh0 = a.shape[0]
+        for sh in a.shape[1:]:
+            if sh != sh0:
+                raise ValueError(
+                    "All dimensions of input must be of equal length"
+                )
+        step = sum(sh0**x for x in range(a.ndim))
+
+    # TODO: implement flatiter for slice key
+    # a.flat[:end:step] = val
+    fl = a.flat
+    if dpnp.isscalar(val):
```

Comment:
We can use `diagonal` function only for 2D arrays with `wrap=False`. We still need a loop (or implement `flatiter` for slice key) for other cases

Reply:
Can we have something like:
```python
    a_sh = a.shape
    tmp_a = dpnp.ravel(a)
    if not dpnp.isscalar(val):
        val = dpnp.ravel(val)
    tmp_a[:end:step] = val
    tmp_a = dpnp.reshape(tmp_a, a_sh)
    a[:] = tmp_a
```
Would it be more performance efficient rather than iterating over the loop?

Reply:
> We can use `diagonal` function only for 2D arrays with `wrap=False`

Btw, `diagonal` must return not writable view. I will prepare a PR to fix that.

Reply:
> We can use `diagonal` function only for 2D arrays with `wrap=False`

Also for wide 2D array, i.e. where `a.shape[0] <= a.shape[1]`.